### ISSUE TITLE
Paragraphs and headings not affected by font-size

### ISSUE THREAD

**Issue by danielweck** [readium/readium-css#112]
Issue originally reported by @jonaslil :
https://github.com/edrlab/thorium-reader/issues/1550

## Summary
All p elements are displayed with the same font size, regardless of publisher css.
## Problem
Lack of variation in font size distorts the layout of title pages and other special text elements. Endnotes, bibliographies, figcaptions containing p:s, which are usually styled smaller, are rendered with body text size (= 1 em).

Variations in font size are a basic typographic means to communicate text structure and hierarchy.

Tested with v1.7.2 on Windows.
### Title page in Thorium
Main title much too small, publisher info much too large:
![Thorium_title](https://user-images.githubusercontent.com/59641495/134168814-85e38f39-c304-4494-b1c6-4f3bd0b122ca.JPG)
### Title page in ADE
![ADE_2](https://user-images.githubusercontent.com/59641495/134171315-ebdfdac7-d783-40ff-9826-3e8d21ce67ef.JPG)
### Sample book
[Download book here](https://www.sls.fi/sv/utgivning/en-tur-om-hornet)
#### HTML
        <section epub:type="titlepage" class="titlepage">
            <p class="book-author">Guss Mattsson</p>
            <p class="book-title">En tur om hörnet</p>
            <p class="book-subtitle">Bref till Dagens Press af G.&#8239;M.</p>
            <p class="editors">Utg. Clas Zilliacus</p>
            <p class="publisher">Svenska litteratursällskapet i Finland <br />Helsingfors 2021</p>
        </section>
#### CSS
p.book-author {
    font-size: 1.4em;
    font-style: italic;
    line-height: 1;
    margin-top: 2em;
    text-indent: 0;
	}
p.book-title {
    font-family: "Source Serif Subhead", serif;
    text-transform: uppercase;
    font-size: 2.2em;
    letter-spacing: 0.025em;
    font-weight: bold;
    line-height: 1;
    margin-top: 1.5em;
    text-indent: 0; 
	}
p.book-author + p.book-title {
    margin-top: 1.5rem;
}
p.book-subtitle {
	font-size: 1.4em;
	line-height: 1.3;
	margin-top: 1.2rem;
	text-indent: 0;
    }
p.editors {
    font-size: 1.25em;
    line-height: 1.3;
    margin-top: 3rem;
    font-style: italic;
   text-indent: 0;
}
p.publisher {
    font-size: 0.9em;
    line-height: 1.4;
    margin-top: 4rem;
    text-indent: 0;
    font-variant: small-caps;
    letter-spacing: 0.08em;
    page-break-before: avoid;
}

________

Issue originally reported by @apgrover :
https://github.com/edrlab/thorium-reader/issues/1508

## Summary
Setting font-size on h1-h6 tags does not render a visual change. Default h1-h6 font sizes appear instead.
## Scope
Observed in Thorium for PC/Mac v1.6.0
## Expected
Setting font-size: 2.5em on h1-h6 will set each element to the same font size.
<img src="https://user-images.githubusercontent.com/14042519/122102224-36777200-cde3-11eb-8748-aac56ce3cc7f.png" width=70% />

## Observed
font-size: 2.5em is ignored on h1-h6 and default font size appears instead.
<img src="https://user-images.githubusercontent.com/14042519/122102242-3bd4bc80-cde3-11eb-851e-047fba11ab13.png" width=70% />

### HTML
```
			<h1 class="gx">A</h1>
			<h2 class="gx">A</h2>
			<h3 class="gx">A</h3>
			<h4 class="gx">A</h4>
			<h5 class="gx">A</h5>
			<h6 class="gx">A</h6>
```

### CSS
```
.gx {
    font-size: 2.5em;
}
```

Sample file available upon request.

**Comment by danielweck** [comment]
> There is another normalization issue which I think would be important to adress, the normalization of heading sizes

Note that I have now merged issue https://github.com/edrlab/thorium-reader/issues/1508 here.

**Comment by danielweck** [comment]
Related / duplicate issue? https://github.com/readium/readium-css/issues/111

**Comment by danielweck** [comment]
Reported in Thorium by @dualpixel-digital :

Hi,

I have one more improvement/fix suggestion for the next release.

I have an alpha version of version 2, with debug option.

I would like to understand why Thorium doesn't respect the font-size value, in space in the h1-h6 tags?

Note the difference between the desired result (Readium) and the result of Thorium

Is there a best formatting practice in CSS to ensure text sizes are recognized?

![thorium-Title](https://user-images.githubusercontent.com/55760305/165653241-b25a0160-39d5-4337-b761-a370b64f641e.jpg)

![readium-Title](https://user-images.githubusercontent.com/55760305/165653244-932f02f7-274f-4764-8578-f83dfeab869d.jpg)

![DebugMode-Thorium](https://user-images.githubusercontent.com/55760305/165653245-33a0c537-04d7-4255-86d3-c91c297e2965.jpg)

**Comment by danielweck** [comment]
Another related issue from Thorium:

https://github.com/edrlab/thorium-reader/issues/1695

Reported by @spudthebud

===>

This is a fictitious user story to illustrate the point

**User story**
I am 3rd year undergraduate with dyslexia. For my assignment, I am getting articles from peer-reviewed journals on a topic.  I found an article that was published in EPUB.  I opened it in Thorium and adjust the font in the settings. However, it does not apply consistently to the abstract or the headings of the article, as shown in this image. It also doesn't apply to fonts in tables.  

![Thorium and font, version 2](https://user-images.githubusercontent.com/8799391/170871770-1bcd3571-ee0a-4add-a07a-fc4d1a044c3e.jpg)

This also applies to an book from a university publisher, as the next image shows:
![ebook, headings and body, font](https://user-images.githubusercontent.com/8799391/170871929-42e2df9d-c907-4699-8b67-d07a7c88c685.jpg)

**Comment by danielweck** [comment]
Merged from https://github.com/edrlab/thorium-reader/issues/1106

It seems I have a similar problem:
I read books that contain code snippets in a monospace font. In order to still be able to read something, I have to greatly enlarge the font, but the code display remains very small in places. Also, the contrast in these areas is too weak. Changing the font and size in the settings does not change anything.

![image](https://user-images.githubusercontent.com/67079945/129332682-049f6f07-83ed-48f3-a922-3aeb3524173b.png)

(Thorium-Reader on Windows 10, epub)

Even though I can read the "normal" text with the screen reader, it is better understanding to read the code examples myself.

Helpful would be therefore:
- a possibility to increase the font size of code examples.
- a better contrast for the areas

But otherwise this is a really nice program, thanks!

**Comment by danielweck** [comment]
Merged from https://github.com/edrlab/thorium-reader/issues/1106

Reading some programming books in EPUB format, I thought it would be cool to render code snippets, that are styled in CSS accordingly, in monospace font.

![Annotation 2020-06-29 103043](https://user-images.githubusercontent.com/62437174/85985837-fb263c80-b9f3-11ea-95df-db3113f3287d.jpg)

@danielweck, sorry, looks like I have instinctively switched the font because default one looks so  Windows9x-ish and it's hard to read, I honestly don't remember when I did this 🙄
And default monospace is almost disappearing on 1920x1080 display.
It wood be so cool if it could override fonts by category, e.g. 'Humanist' for sans serif, 'Monospace' for monospace etc.

![image](https://user-images.githubusercontent.com/62437174/86084905-9e378e80-baa6-11ea-8d33-05a95b869d90.png)

**Comment by danielweck** [comment]
> this sounds like something that is controlled by https://github.com/readium/readium-css right? 

Indeed but we’re only applying `font-family` to [a subset of HTML elements](https://github.com/readium/readium-css/blob/master/css/src/modules/user-settings-submodules/ReadiumCSS-fontFamily_pref.css), which doesn’t contain markup elements for code, etc. 

So it seems to me the user setting is applied to code because it’s using `p class` in this publication. If the markup in the book was using `pre + code` it shouldn’t have been overridden.

That said, it wouldn’t solve this rendering issue raised by @weiss-d 

> And default monospace is almost disappearing on 1920x1080 display. It wood be so cool if it could override fonts by category, e.g. 'Humanist' for sans serif, 'Monospace' for monospace etc.

That’s quite tricky with CSS only since emulating defaults (for `sans-serif`, `serif`, `monospace`, `script`, etc.) is not really a thing it was meant for. It’s more of an app concern as they have easier access to that (+ installed fonts) and then there’s the issue of force-overriding the authors’ stylesheets (cf. browsers). 

![Capture d’écran 2020-06-30 à 10 49 21](https://user-images.githubusercontent.com/12599652/86105302-730d6900-babf-11ea-857e-caa81239c2f7.png)

But I can imagine Electron has something for that?

**Comment by danielweck** [comment]
Hello @man-we , thank you for your feedback.

Readium CSS is an open-source project that could do with some love from developers / maintainers.
Would your employer ("large publisher") be able to offer some help?

In Thorium we implemented a few tweaks / workarounds in order to address weird edge-cases in ReadiumCSS, but we always report problems upstream (here in this repository) just in case the issues we encounter specifically in Electron/Chromium also generally affect other projects (e.g. the Readium "mobile" iOS and Android toolkits).

Besides exploring issues related to proportional font scaling, I have been looking at ways to disable some ReadiumCSS features in order to come as close as possible to "leave me alone" (i.e. strict minimum alterations of publisher styles in order to preserve "essential" reading-system overrides, such as colour themes, word/line/paragraph spacing, etc.). However, whenever I turn off some of the "normalised" styling, I seem to come across weird edge-case issues with image dimensions / scaling (especially in paginated / CSS columns mode, not so much in vertical scrolling mode), or internationalisation problems (locales that have specific glyph rendering rules) ... the original developer of ReadiumCSS did a lot of research in order to come up with a sane baseline.

All that said, I completely agree that this particular font sizing issue needs to be bumped up the list of priorities. Let me have a look at Thorium/Electron/Chromium specifically. I need to revisit the workarounds I tried in past iterations.

**Comment by danielweck** [comment]
report from @CLnordcompo

Here is the problem:
Thorium poorly manages the size of titles and intertitles.
It supports h1, h2, h3 tags and assigns a default size, depending on the size configured in user style sheets.
This is not the case for other e-readers (ADE, iPad, etc.).
Often implying that the same box title does not have the same rendering on the same work, the size depends on the level to which the box belongs.
I am attaching a snippet of epub, with corresponding CSS, illustrating the case.
[Test.zip](https://github.com/edrlab/thorium-reader/files/14633972/Test.zip)

Thanks in advance

**Comment by danielweck** [comment]
From https://github.com/edrlab/thorium-reader/issues/2073
By @riiski

Hi,
I have noticed when changing reading fonts in Reader app on Windows that fonts does not change header font. Only "Duospace" font change headers (H1, h2, h3, ...).
In this pic you can see that paragraph font is changed while header font is the same:
![2024-03-05 10_59_27-Thorium - Balansas  Skanus maistas ir sveikos gyvensenos idėjos aktyviems žmonėm](https://github.com/edrlab/thorium-reader/assets/7087900/e0dc98ab-6938-4c8d-9ce6-097f8d57e545)

The code clearly shows that it is h1 header tag on xhtml file:
![2024-03-05 11_00_01-sfr2_balansas_v1 epub - epub3 0 - Sigil](https://github.com/edrlab/thorium-reader/assets/7087900/38d20d24-dc8b-4629-8a41-d02088bf759a)

And css values on the h1 tag:
![2024-03-05 11_00_19-sfr2_balansas_v1 epub - epub3 0 - Sigil](https://github.com/edrlab/thorium-reader/assets/7087900/5ea3cd25-8ca7-41af-9578-be93d1830577)

**Comment by jonaslil** [comment]
The two main problems discussed in this issue remain unresolved in Thorium: font size settings in the EPUB’s own CSS for 1) paragraphs and 2) headings (h1–h6) have no effect. However, the issue is marked as completed. Does this mean improvements were made in Readium CSS but have not yet been implemented in Thorium?

This issue was recently noted in CSS development efforts at Nordic special libraries (MTM, Celia), where ensuring proper handling of EPUB styles is important for accessibility.

Tested in Thorium 3.0.0 (Windows) and Thorium 3.1.0 (latest test build for macOS Intel).

**Comment by danielweck** [comment]
Hello, Readium CSS v2 is now available but as far as I know is not integrated in any production Readium reading system yet (mobile iOS / Android, or desktop Thorium). I am not sure about the ongoing web toolkit / demo app. @JayPanoz ?

**Comment by HadrienGardeur** [comment]
Readium CSS 2.0 is used in Thorium Web but hasn't been integrated in other toolkits yet.

Our next iteration in February on the Preferences API should help future adoption in Readium Mobile down the road.

**Comment by JayPanoz** [comment]
@danielweck TS-Toolkit is still shipping with version 1 due to the lack of Preferences API. We will start implementing the API in February with an ETA in early March. This new major version will ship with ReadiumCSS v2. 

Readium Playground/Thorium Web is already using ReadiumCSS v2 through a specific submodule as we needed it to make some early progress on `auto` pagination, multiple columns, etc. and improve other parts e.g. theming. 

The lack of the Preferences API is a major pain point as we need to use methods that are not meant to be public and update everything manually though, which is why we will switch focus to the Ts-toolkit, and backport auto-pagination in the Ts-Toolkit itself so that FXL and reflow can be consistent.

**Comment by jonaslil** [comment]
Thanks for the update! Just to confirm, once the Preferences API is implemented, will the next Thorium Desktop release include ReadiumCSS v2, or is there additional work required before that happens? Also, will this update resolve the issues I mentioned earlier, where font size settings in the EPUB’s own CSS for paragraphs and headings have no effect?

**Comment by JayPanoz** [comment]
Thorium Desktop is not using the ts-toolkit’s Navigator so they don’t have to wait for the Preference API. Note Readium CSS v2 has some significant breaking changes since it follows a different philosophy though, which is why it’s not necessarily trivial to upgrade. A lot of developer pain points have been removed but it also means some features ReadiumCSS used to provide have to be re-implemented. 

The font-size user setting has been entirely rewritten in version 2 and is now entirely delegating the feature to the browser. It should be consistent with what browsers do on `Ctrl/Cmd +` and `Ctrl/Cmd -`.

**Comment by jonaslil** [comment]
Thanks for the explanation! Just to clarify, does Readium CSS v2 now respect the font-size settings defined in the EPUB’s own CSS for paragraphs and headings, or are those overridden by default?

**Comment by JayPanoz** [comment]
@jonaslil v2 is not targeting any element for the font-size anymore, it's just applying a zoom factor globally. So I confirm it leaves headings and paragraphs untouched. 

The normalisation that was done in v1 was a really difficult decision to make at the time since applying a font-size for `:root` would have made the user setting broken for something like 20–25% of the EPUBs we analyzed at the time. It was primarily motivated by accessibility, to make problematic publications readable, albeit  with suboptimal handling.

There were discussions on alternative options but nobody wanted to go back to the Readium.js implementation in JS due to performance concerns (getting the computed value for every element in the DOM and applying a factor on font-size change is very costly). 

As a matter of fact, one of the goals of V2 when I got back into the maintenance was to solve this issue specifically.
